### PR TITLE
[4033] Extract user-scope from the existing client-scope

### PR DESCRIPTION
**Pull request by kanat**
### 🎯 Goal

Extract user-scope from the existing client-scope

### 🛠 Implementation details

instead of single `scope` ChatClient now has 2 scopes: `clientScope` & `userScope` where `clientScope` is used for `connectAnyUser` & `disconnect` functions

### 🧪 Testing

- Run any Sample app
- Login
- Validate base features, like sending msg, receiving msg etc

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: stream-chat-android-client/src/test/java/io/getstream/chat/android/client/socket/HealthMonitorTest.kt [hunk 1]
```diff
@@ -32,16 +38,32 @@ import org.mockito.kotlin.whenever
...
+    @BeforeEach
+    fun setUp() {
+        StreamLog.setValidator { _, _ -> true }
+        StreamLog.setLogger(
+            KotlinStreamLogger(now = {
+                testCoroutines.dispatcher.scheduler.currentTime
+            })
+        )
+    }
+
```
Comment (JcMinarro) [inline stream-chat-android-client/src/test/java/io/getstream/chat/android/client/socket/HealthMonitorTest.kt:56]: Can we create a helper method to init logger?
It seems to be repeated across our tests

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/configuration/Config.kt [hunk 2]
```diff
@@ -32,5 +32,5 @@ public data class Config @JvmOverloads constructor(
     public val userPresence: Boolean = true,
     public val persistenceEnabled: Boolean = true,
     public val uploadAttachmentsNetworkType: UploadAttachmentsNetworkType = UploadAttachmentsNetworkType.NOT_ROAMING,
-    public val useSequentialEventHandler: Boolean = false,
+    public val useSequentialEventHandler: Boolean = true,
```
Comment (JcMinarro) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/configuration/Config.kt:35]: Why this change was needed?

File: stream-chat-android-test/src/main/java/io/getstream/chat/android/test/TestCoroutineExtension.kt [hunk 3]
```diff
@@ -20,6 +20,7 @@ package io.getstream.chat.android.test
 
 import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.plus
```
Comment (JcMinarro) [inline stream-chat-android-test/src/main/java/io/getstream/chat/android/test/TestCoroutineExtension.kt:23]: This import seems that is not needed anymore

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 4]
```diff
@@ -243,15 +246,24 @@ internal constructor(
...
     }
 
+    internal fun launch(
+        block: suspend CoroutineScope.() -> Unit
+    ) = userScope.launch(block = block)
+
+    @InternalStreamChatApi
+    public fun inheritScope(block: (Job) -> CoroutineContext): CoroutineScope {
+        return userScope + block(userScope.coroutineContext.job)
+    }
```
Comment (JcMinarro) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:256]: Can we add docs to this method?
When should we use it?
Why use it and not the `userScope` directly?

**Review by JcMinarro — COMMENTED** [review]
Great job!!!!
Overall it looks good, I left some comments

File: stream-chat-android-client/src/test/java/io/getstream/chat/android/client/ChatClientTest.kt [hunk 5]
```diff
@@ -261,6 +274,7 @@ internal class ChatClientTest {
 
         /* Then */
         result shouldBeEqualTo Result.error(ChatError("channelsIds must contain at least 1 id."))
+        println("!.!.!")
```
[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-client/src/test/java/io/getstream/chat/android/client/ChatClientTest.kt:277]: Good catch man!

There was a problem with not ending tests, so I was investigating that problem and forgot to delete this output 🙈

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/scope/ClientScope.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import io.getstream.chat.android.client.call.SharedCalls
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+
+internal interface ClientScope : CoroutineScope
+
+internal fun ClientScope(): ClientScope = ClientScopeImpl()
+
+private class ClientScopeImpl :
```
[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/scope/ClientScope.kt:28]: There is a slight diff implementation for unit tests called `ClientTestScope` which extends the `testScope`.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/scope/UserScope.kt [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.job
+import kotlinx.coroutines.plus
+
+internal interface UserScope : CoroutineScope
+
+internal fun UserScope(clientScope: ClientScope): UserScope = UserScopeImpl(clientScope)
+
+private class UserScopeImpl(
```
[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/scope/UserScope.kt:28]: same here we've got `UserTestScope` for unit tests

File: stream-chat-android-client/src/test/java/io/getstream/chat/android/client/socket/HealthMonitorTest.kt (hunk 1 above)
Comment (kanat) [reply stream-chat-android-client/src/test/java/io/getstream/chat/android/client/socket/HealthMonitorTest.kt:56]: Makes sense 👌

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/configuration/Config.kt (hunk 2 above)
Comment (kanat) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/plugin/configuration/Config.kt:35]: Yeah, it is not a part of this PR, I just noticed this change is missing in the following PRs:
- https://github.com/GetStream/stream-chat-android/pull/4083
- #4089

**Review by JcMinarro — APPROVED** [review]
